### PR TITLE
Merge spark components with pandas-profiling develop

### PR DESCRIPTION
**Pull request by chanedwin**

### CODE REVIEW

File: src/pandas_profiling/model/dataframe_wrappers.py [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+
+    def dropna(self, subset) -> "PandasDataFrame":
+        return PandasDataFrame(self.df.dropna(subset=subset))
+
+    def groupby_get_n_largest_dups(
+        self, supported_columns, n_head
+    ) -> Tuple[Dict[str, Any], Optional[pd.DataFrame]]:
+
+        metrics: Dict[str, Any] = {}
+        if n_head > 0:
```
Comment (ieaves) [inline src/pandas_profiling/model/dataframe_wrappers.py:264]: I would consider re-writing this function just to reduce the level of nesting e.g.

```
if n_head <= 0:
  return metrics, None

if not supported_columns or len(self.df) == 0:
  metrics["n_duplicates"] = 0
  metrics["p_duplicates"] = 0.0
  return metrics, None

// remainder of un-nested implementation
```

Comment (ieaves) [reply src/pandas_profiling/model/dataframe_wrappers.py:264]: Also, that can just be `self.df.empty`

File: src/pandas_profiling/model/dataframe_wrappers.py [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+        """
+
+        Args:
+            subset:
+            keep:
+
+        Returns:
+
+        """
+        return len(self.df[self.df.duplicated(subset=subset, keep="first")])
```
Comment (ieaves) [inline src/pandas_profiling/model/dataframe_wrappers.py:254]: You can avoid the index & copy here

`return self.df.duplicate(subset=subset).sum()`

File: src/pandas_profiling/model/dataframe_wrappers.py [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+
+        Args:
+            df: the pandas DataFrame
+
+        Returns:
+            The preprocessed DataFrame
+        """
+
+        # Treat index as any other column
+        if (
```
Comment (ieaves) [inline src/pandas_profiling/model/dataframe_wrappers.py:223]: I think this is getting at whether the index is a `RangeIndex` with start=0, and step=1 (the default index type) which can be checked without building a new index and comparing all values directly. e.g.

```
if not isinstance(df.index, pd.RangeIndex) or not (df.index.start == 0 and df.index.step == 1):
  df = df.reset_index()
```

That conditional looks a little weird but it's equivalent to `not (isinstance(df.index, pd.RangeIndex) and (df.index.start == 0 and df.index.step == 1))` via De Morgan's law.
```

**Comment by Baukebrenninkmeijer** [comment]
@chanedwin Just wanted to say that this is a great addition and I'm really glad it's being added!

**Comment by Baukebrenninkmeijer** [comment]
@chanedwin I'm running into some errors when running this. 

```python
ProfileReport(sdf, minimal=True)
```
```
KeyError                                  Traceback (most recent call last)
/databricks/python/lib/python3.7/site-packages/IPython/core/formatters.py in __call__(self, obj)
    343             method = get_real_method(obj, self.print_method)
    344             if method is not None:
--> 345                 return method()
    346             return None
    347         else:

/databricks/python/lib/python3.7/site-packages/pandas_profiling/profile_report.py in _repr_html_(self)
    443     def _repr_html_(self):
    444         """The ipython notebook widgets user interface gets called by the jupyter notebook."""
--> 445         self.to_notebook_iframe()
    446 
    447     def __repr__(self):

/databricks/python/lib/python3.7/site-packages/pandas_profiling/profile_report.py in to_notebook_iframe(self)
    423         with warnings.catch_warnings():
    424             warnings.simplefilter("ignore")
--> 425             display(get_notebook_iframe(self))
    426 
    427     def to_widgets(self):

/databricks/python/lib/python3.7/site-packages/pandas_profiling/report/presentation/flavours/widget/notebook.py in get_notebook_iframe(profile)
     63         output = get_notebook_iframe_src(profile)
     64     elif attribute == "srcdoc":
---> 65         output = get_notebook_iframe_srcdoc(profile)
     66     else:
     67         raise ValueError(

/databricks/python/lib/python3.7/site-packages/pandas_profiling/report/presentation/flavours/widget/notebook.py in get_notebook_iframe_srcdoc(profile)
     21     width = config["notebook"]["iframe"]["width"].get(str)
     22     height = config["notebook"]["iframe"]["height"].get(str)
---> 23     src = html.escape(profile.to_html())
     24 
     25     iframe = f'<iframe width="{width}" height="{height}" srcdoc="{src}" frameborder="0" allowfullscreen></iframe>'

/databricks/python/lib/python3.7/site-packages/pandas_profiling/profile_report.py in to_html(self)
    393 
    394         """
--> 395         return self.html
    396 
    397     def to_json(self) -> str:

/databricks/python/lib/python3.7/site-packages/pandas_profiling/profile_report.py in html(self)
    208     def html(self):
    209         if self._html is None:
--> 210             self._html = self._render_html()
    211         return self._html
    212 

/databricks/python/lib/python3.7/site-packages/pandas_profiling/profile_report.py in _render_html(self)
    315         from pandas_profiling.report.presentation.flavours import HTMLReport
    316 
--> 317         report = self.report
    318 
    319         disable_progress_bar = not config["progress_bar"].get(bool)

/databricks/python/lib/python3.7/site-packages/pandas_profiling/profile_report.py in report(self)
    202     def report(self):
    203         if self._report is None:
--> 204             self._report = get_report_structure(self.description_set)
    205         return self._report
    206 

/databricks/python/lib/python3.7/site-packages/pandas_profiling/profile_report.py in description_set(self)
    182         if self._description_set is None:
    183             self._description_set = describe_df(
--> 184                 self.title, self.df, self.summarizer, self.typeset, self._sample
    185             )
    186         return self._description_set

/databricks/python/lib/python3.7/site-packages/pandas_profiling/model/describe.py in describe(title, df, summarizer, typeset, sample)
    167         # Messages
    168         pbar.set_postfix_str("Get messages/warnings")
--> 169         messages = get_messages(table_stats, series_description, correlations)
    170         pbar.update()
    171 

/databricks/python/lib/python3.7/site-packages/pandas_profiling/model/messages.py in get_messages(table_stats, series_description, correlations)
    328     messages = check_table_messages(table_stats)
    329     for col, description in series_description.items():
--> 330         messages += check_variable_messages(col, description)
    331     messages += check_correlation_messages(correlations)
    332     messages.sort(key=lambda message: str(message.message_type))

/databricks/python/lib/python3.7/site-packages/pandas_profiling/model/messages.py in check_variable_messages(col, description)
    293         messages += unsupported_warnings(description)
    294     else:
--> 295         messages += supported_warnings(description)
    296 
    297         if description["type"] == Categorical:

/databricks/python/lib/python3.7/site-packages/pandas_profiling/model/messages.py in supported_warnings(summary)
    238         bool
    239     ):
--> 240         if summary["n_distinct"] == summary["n"]:
    241             messages.append(
    242                 Message(

KeyError: 'n_distinct'
```

And Nr. 2 for the non-minimal:

```python
ProfileReport(sdf)
```

```
ZeroDivisionError                         Traceback (most recent call last)
/databricks/python/lib/python3.7/site-packages/IPython/core/formatters.py in __call__(self, obj)
    343             method = get_real_method(obj, self.print_method)
    344             if method is not None:
--> 345                 return method()
    346             return None
    347         else:

/databricks/python/lib/python3.7/site-packages/pandas_profiling/profile_report.py in _repr_html_(self)
    443     def _repr_html_(self):
    444         """The ipython notebook widgets user interface gets called by the jupyter notebook."""
--> 445         self.to_notebook_iframe()
    446 
    447     def __repr__(self):

/databricks/python/lib/python3.7/site-packages/pandas_profiling/profile_report.py in to_notebook_iframe(self)
    423         with warnings.catch_warnings():
    424             warnings.simplefilter("ignore")
--> 425             display(get_notebook_iframe(self))
    426 
    427     def to_widgets(self):

/databricks/python/lib/python3.7/site-packages/pandas_profiling/report/presentation/flavours/widget/notebook.py in get_notebook_iframe(profile)
     63         output = get_notebook_iframe_src(profile)
     64     elif attribute == "srcdoc":
---> 65         output = get_notebook_iframe_srcdoc(profile)
     66     else:
     67         raise ValueError(

/databricks/python/lib/python3.7/site-packages/pandas_profiling/report/presentation/flavours/widget/notebook.py in get_notebook_iframe_srcdoc(profile)
     21     width = config["notebook"]["iframe"]["width"].get(str)
     22     height = config["notebook"]["iframe"]["height"].get(str)
---> 23     src = html.escape(profile.to_html())
     24 
     25     iframe = f'<iframe width="{width}" height="{height}" srcdoc="{src}" frameborder="0" allowfullscreen></iframe>'

/databricks/python/lib/python3.7/site-packages/pandas_profiling/profile_report.py in to_html(self)
    393 
    394         """
--> 395         return self.html
    396 
    397     def to_json(self) -> str:

/databricks/python/lib/python3.7/site-packages/pandas_profiling/profile_report.py in html(self)
    208     def html(self):
    209         if self._html is None:
--> 210             self._html = self._render_html()
    211         return self._html
    212 

/databricks/python/lib/python3.7/site-packages/pandas_profiling/profile_report.py in _render_html(self)
    315         from pandas_profiling.report.presentation.flavours import HTMLReport
    316 
--> 317         report = self.report
    318 
    319         disable_progress_bar = not config["progress_bar"].get(bool)

/databricks/python/lib/python3.7/site-packages/pandas_profiling/profile_report.py in report(self)
    202     def report(self):
    203         if self._report is None:
--> 204             self._report = get_report_structure(self.description_set)
    205         return self._report
    206 

/databricks/python/lib/python3.7/site-packages/pandas_profiling/profile_report.py in description_set(self)
    182         if self._description_set is None:
    183             self._description_set = describe_df(
--> 184                 self.title, self.df, self.summarizer, self.typeset, self._sample
    185             )
    186         return self._description_set

/databricks/python/lib/python3.7/site-packages/pandas_profiling/model/describe.py in describe(title, df, summarizer, typeset, sample)
     92         total=number_of_tasks, desc="Summarize dataset", disable=disable_progress_bar
     93     ) as pbar:
---> 94         series_description = get_series_descriptions(df, summarizer, typeset, pbar)
     95 
     96         pbar.set_postfix_str("Get variable types")

/databricks/python/lib/python3.7/functools.py in wrapper(*args, **kw)
    838                             '1 positional argument')
    839 
--> 840         return dispatch(args[0].__class__)(*args, **kw)
    841 
    842     funcname = getattr(func, '__name__', 'singledispatch function')

/databricks/python/lib/python3.7/site-packages/pandas_profiling/model/summary_spark.py in get_series_descriptions_spark(df, summarizer, typeset, pbar)
     91             SparkDataFrame(df.get_spark_df().select(list(unsupported_cols))),
     92             engine=df.engine,
---> 93             dtype=SparkUnsupported,
     94         )
     95     )

/databricks/python/lib/python3.7/site-packages/pandas_profiling/model/summarizer.py in summarize(self, data, engine, dtype)
     63         elif engine == "spark":
     64             _, summary = summarizer_func(
---> 65                 data, {column: {"type": dtype} for column in data.columns}
     66             )
     67         else:

/databricks/python/lib/python3.7/site-packages/pandas_profiling/model/handler.py in func2(*x)
     21                 return f(*x)
     22             else:
---> 23                 return f(*res)
     24 
     25         return func2

/databricks/python/lib/python3.7/site-packages/pandas_profiling/model/handler.py in func2(*x)
     21                 return f(*x)
     22             else:
---> 23                 return f(*res)
     24 
     25         return func2

/databricks/python/lib/python3.7/site-packages/pandas_profiling/model/handler.py in func2(*x)
     17     def func(f, g):
     18         def func2(*x):
---> 19             res = g(*x)
     20             if type(res) == bool:
     21                 return f(*x)

/databricks/python/lib/python3.7/functools.py in wrapper(*args, **kw)
    838                             '1 positional argument')
    839 
--> 840         return dispatch(args[0].__class__)(*args, **kw)
    841 
    842     funcname = getattr(func, '__name__', 'singledispatch function')

/databricks/python/lib/python3.7/site-packages/pandas_profiling/model/summary_algorithms_spark.py in describe_supported_spark(df, series_description)
    128             distinct_count = series.distinct
    129             stats.update(
--> 130                 {"n_distinct": distinct_count, "p_distinct": distinct_count / count}
    131             )
    132         if config["spark"]["compute_unique"].get(bool):

ZeroDivisionError: division by zero
```

I cannot share too much about the data, but I can share these statistics. These were generated with the non-spark pandas profiling. I think both errors are related with empty numerical columns, but I'm running this in databricks so I can't easily troubleshoot there. 
```python
{
    "n": 2955,
    "n_var": 69,
    "memory_size": 1631288,
    "record_size": 552.0433164128596,
    "n_cells_missing": 17730,
    "n_vars_with_missing": 6,
    "n_vars_all_missing": 6,
    "p_cells_missing": 0.08695652173913043,
    "types": {
        "Categorical": 48,
        "Unsupported": 19,
        "DateTime": 2
	}
}
```

**Comment by sbrugman** [comment]
Hi @Baukebrenninkmeijer, this branch is outdated but spark support will be added soon ;)!

**Comment by Baukebrenninkmeijer** [comment]
@sbrugman Oh, which one is the correct one?  I did see several referenced from several issues, but this was the only open PR relating to this.

**Comment by mkopec87** [comment]
> Hi @Baukebrenninkmeijer, this branch is outdated but spark support will be added soon ;)!

Is there a possibility to test the spark support already? Any development branch or sth? :)

**Comment by chanedwin** [comment]
Closing as this is outdated.

**Pull request closed without merging**